Added map to the app

IanisT227 opened the pull request:
Added the map with only the first cardView that shows scooter info. I wanted to check if I implemented this feature in the proper manner before carrying on with the task

IanisT227 committed:
Added map on MapFragment and put a marker in Galati 😩 to check if it works

IanisT227 committed:
Added all scooters on a 4KM radius as markers on map

IanisT227 committed:
Refactoring, a lot of it

IanisT227 committed:
Implemented Scooter CardView

IanisT227 committed:
Implemented Scooter CardView

IanisT227 committed:
Changed some code according to the PR comments, still have a few issues to address

IanisT227 committed:
Solved a bug that allowed the user to click the map on the area of the custom cardview

IanisT227 committed:
passing token via interceptor works now

IanisT227 committed:
Added an error message if the upload of driving license fails. Initialized some views directly in the xml instead of doing a method only for that in the fragment.

IanisT227 committed:
Added an error message if the upload of driving license fails. Initialized some views directly in the xml instead of doing a method only for that in the fragment. + some import optimizing

IanisT227 committed:
Added animations on mapFragment

IanisT227 committed:
Added battery level indicator on mapFragment
reorganized some strings

IanisT227 commented:
I've also solved the issues from the previous PR here

IanisT227 committed:
Added double press to exit the app

IanisT227 committed:
Added some map styling
no cluster manager yet

catadarjan-tapptitude commented on app/build.gradle:
```diff
@@ -49,29 +50,43 @@ android {
+
+    //FARA ASTEA CRAPA TOT ASTEA SUNT DE LA ANDROID
+    implementation "androidx.fragment:fragment-ktx:1.3.3"
+    implementation 'com.google.code.gson:gson:2.8.8'
```
Do you still need Gson?

catadarjan-tapptitude commented on app/build.gradle:
```diff
@@ -49,29 +50,43 @@ android {
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
-    implementation("androidx.navigation:navigation-fragment-ktx:2.5.1")
-    implementation("androidx.navigation:navigation-ui-ktx:2.5.1")
+
```
Extra empty line

catadarjan-tapptitude commented on app/build.gradle:
```diff
@@ -49,29 +50,43 @@ android {
+    implementation("androidx.navigation:navigation-fragment-ktx:2.5.1")
+    implementation("androidx.navigation:navigation-ui-ktx:2.5.1")
+
+    //FARA ASTEA CRAPA TOT ASTEA SUNT DE LA ANDROID
```
I suggest using all English and also avoid using all caps :P 
Also, this is something I forgot to mention:
Use the region/endregion comment tags keywords.
This lets you collapse regions. It's a cool trick
![image](https://user-images.githubusercontent.com/84084708/191166059-d41e18d8-c1fc-4bcb-9ebd-68a427888a1b.png)

![image](https://user-images.githubusercontent.com/84084708/191166049-0b7c1abe-d4c0-4aaf-8bbe-efde4fd02764.png)

catadarjan-tapptitude commented on app/build.gradle:
```diff
@@ -49,29 +50,43 @@ android {
+    implementation 'androidx.appcompat:appcompat:1.5.0'
+
+    //TEST
     testImplementation 'junit:junit:4.13.2'
```
Since we aren't going to implement tests, you can remove these dependencies and the example test files/folders

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/MainActivity.kt:
```diff
@@ -22,4 +28,24 @@ class MainActivity : AppCompatActivity(), OnSkipButtonPressed {
         findNavController(R.id.navigationHost).navigate(SplashFragmentDirections.actionGlobalRegisterFragment())
     }
+
+    fun View.hideKeyboard() {
```
This extension function should either be static private or extracted into a Views extension file

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/MainActivity.kt:
```diff
@@ -22,4 +28,24 @@ class MainActivity : AppCompatActivity(), OnSkipButtonPressed {
     }
+
+    fun View.hideKeyboard() {
+        val inputManager = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
```
Since you're in an activity, you can call `getSystemService` directly w/o `context.`
Import `INPUT_METHOD_SERVICE` directly
This method could be inlined as
`(getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(windowToken, RESULT_UNCHANGED_SHOWN)`
But that's just a subjective preference, up to you :P

When using methods that require a flag parameter, try looking for the constant that represents it in the related class(es). Here for example, 0 is refering to `RESULT_UNCHANGED_SHOWN` from `InputMethodManager`
See here
https://developer.android.com/reference/android/view/inputmethod/InputMethodManager#summary

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/MainActivity.kt:
```diff
@@ -22,4 +28,24 @@ class MainActivity : AppCompatActivity(), OnSkipButtonPressed {
+        inputManager.hideSoftInputFromWindow(windowToken, 0)
+    }
+
+    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
```
What do you need this override for?
Let's discuss in a huddle

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/AuthenticationViewModel.kt:
```diff
@@ -19,30 +19,38 @@ class AuthenticationViewModel(
     val onUserLoggedIn: MutableLiveData<Int> = MutableLiveData(UNCHECKED)
     val userData: MutableLiveData<UserResponse> = MutableLiveData(null)
+    val isLoading: MutableLiveData<Boolean> = MutableLiveData(false)
+
```
Extra empty line

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/AuthenticationViewModel.kt:
```diff
@@ -19,30 +19,38 @@ class AuthenticationViewModel(
 
     val onUserLoggedIn: MutableLiveData<Int> = MutableLiveData(UNCHECKED)
     val userData: MutableLiveData<UserResponse> = MutableLiveData(null)
+    val isLoading: MutableLiveData<Boolean> = MutableLiveData(false)
```
Keep in mind to use backing fields for MutableLiveData, such that it cannot be changed from outside the VM.

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/login/LoginFragment.kt:
```diff
@@ -99,7 +100,22 @@ class LoginFragment : Fragment(R.layout.fragment_login) {
+        }
+
+        viewModel.isLoading.observe(viewLifecycleOwner)
+        { logValue ->
```
Check Android Studio formatting settings, since this `{` should be on the line above.
I'd name this to mirror the liveData name, `isLoading`. Having the `Value` suffix feels a bit redundant since this is a value of a variable. Also, `logValue` doesn't represent what the LiveData variable is, it's related to loading state, not login state

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/login/LoginFragment.kt:
```diff
@@ -99,7 +100,22 @@ class LoginFragment : Fragment(R.layout.fragment_login) {
+
+        viewModel.isLoading.observe(viewLifecycleOwner)
+        { logValue ->
+            binding.apply {
```
Imo there is not much benefit to using `.apply` . Yes, it reduces the amount of time you type `binding` but it also increases indentation of your code. I would err on the side of using `binding.` in this case.
However if you apply multiple changes on a single view, then that'd be more fitting, the way I see it.

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/AuthenticationViewModel.kt:
```diff
@@ -19,30 +19,38 @@ class AuthenticationViewModel(
 
     val onUserLoggedIn: MutableLiveData<Int> = MutableLiveData(UNCHECKED)
```
Let's discuss this approach and the usage of `AuthenticationViewModel` when you have some time

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/register/RegisterFragment.kt:
```diff
@@ -118,8 +125,25 @@ class RegisterFragment : Fragment(R.layout.fragment_register) {
             }
         }
+
+        viewModel.isLoading.observe(viewLifecycleOwner)
```
Same comments as in `LoginFragment`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/authentication/register/RegisterFragment.kt:
```diff
@@ -118,8 +125,25 @@ class RegisterFragment : Fragment(R.layout.fragment_register) {
                 findNavController().navigate(RegisterFragmentDirections.actionRegisterFragmentToLicenseInstructionsFragment(userData = userResponse))
             } else if (logValue == ERROR) {
-                Toast.makeText(context, "Email already exists", Toast.LENGTH_SHORT).show()
+                Alerter.create(requireActivity())
```
You're using this piece of code in multiple places.
It could be extracted in a separate static method, or you could implement and `AlerterHandler` and injecting it where needed. 
Let's discuss about this too

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseConfirmFragment.kt:
```diff
@@ -26,21 +26,24 @@ class LicenseConfirmFragment : Fragment(R.layout.fragment_license_confirm) {
-                binding.validationStatusTV.text = "We’ve succesfuly validated your driving license"
-                binding.imageView.visibility = View.VISIBLE
-                binding.launchMapBtn.visibility = View.VISIBLE
+            binding.apply {
```
Same comments as in `LoginFragment`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseConfirmFragment.kt:
```diff
@@ -26,21 +26,24 @@ class LicenseConfirmFragment : Fragment(R.layout.fragment_license_confirm) {
+                validationStatusTV.isVisible = loadingValue
+                validationConfirmTV.isVisible = !loadingValue
+            }
+            if (licenseRegistrationViewModel.isError.equals(true)) {
```
Using another LiveData value inside another observer isn't by itself a bad thing.
However, the way you use it here could be moved to an observer for `isError` instead.
Let's discuss

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseInstructionsFragment.kt:
```diff
@@ -38,25 +43,32 @@ class LicenseInstructionsFragment : Fragment(R.layout.fragment_license_instructi
+                Toast.makeText(requireContext(), getString(R.string.exit_app_button_text), Toast.LENGTH_SHORT).show()
+                Handler().postDelayed(Runnable {
+                    doubleBackPressed = false
+                }, 3000L)
```
Extract magic numbers to constants

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseInstructionsFragment.kt:
```diff
@@ -67,18 +79,11 @@ class LicenseInstructionsFragment : Fragment(R.layout.fragment_license_instructi
 
     private fun getTmpFileUri(): Uri {
-        val tmpFile = File.createTempFile("tmp_image_file", ".png").apply {
+        val tmpFile = File.createTempFile(getString(R.string.create_image_file_prefix), getString(R.string.create_image_file_sufix)).apply {
```
The strings you extracted aren't displayed to the user. This means that you could have them in the kt companion object instead

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseItem.kt:
```diff
@@ -1,11 +1,11 @@
 data class LicenseItem(
     val token: String,
-    val image: File
+    val imageUri: Uri
```
You could go one step further and have this as a string. Then turn it into a Uri in the receiving fragment.
This will make the parcelize process even faster

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseRegistrationViewModel.kt:
```diff
@@ -12,14 +12,14 @@ import okhttp3.MultipartBody
 ) : ViewModel() {
-
-    val licenseResponse: MutableLiveData<User?> = MutableLiveData()
     val isLoading: MutableLiveData<Boolean> = MutableLiveData(false)
```
Leave an empty line between class declaration and first method/field

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseRegistrationViewModel.kt:
```diff
@@ -12,14 +12,14 @@ import okhttp3.MultipartBody
+        get() = _isError
 
     fun uploadImage(token: String, image: File) {
         viewModelScope.launch {
```
Now that you've implemented the SessionInterceptor, you can remove the `token` field from here. Sinceyou can implement it to be added by the Interceptor for all the necessary API calls

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/Coordinates.kt:
```diff
@@ -0,0 +1,10 @@
+import com.squareup.moshi.JsonClass
+
+@JsonClass(generateAdapter = true)
+data class Coordinates(
```
When it comes to classes used in Network API calls, I would use the suffix `Dto` (or `DTO`) to express the fact that they are being used as Data Transfer Objects. This makes it easier to differentiate their usage from a `Coordinates` class that's used somewhere else in the architecture.
`CoordinatesDto`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapService.kt:
```diff
@@ -0,0 +1,10 @@
+import retrofit2.http.GET
+import retrofit2.http.Query
+
+interface MapService {
```
I think this would be more fittingly named as `ScootersService`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/ScooterResponse.kt:
```diff
@@ -0,0 +1,22 @@
+    @Json(name = "battery")
+    val battery: String,
+    @Json(name = "updatedAt")
+    val updatedAt: String,
```
Do you really need this field?

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapViewModel.kt:
```diff
@@ -0,0 +1,24 @@
+import kotlinx.coroutines.launch
+
+class MapViewModel(private val mapService: MapService) : ViewModel() {
+    val scooterList: MutableLiveData<List<ScooterResponse>> = MutableLiveData()
```
Leave an empty line between class declaration and first field/method

catadarjan-tapptitude replied:
Use LiveData backing fields for MutableLiveData

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapViewModel.kt:
```diff
@@ -0,0 +1,24 @@
+                scooterList.value = mapService.getAllScooters(latitude = latitude, longitude = longitude)
+                logTag("SCOOTER_0", mapService.getAllScooters(latitude = latitude, longitude = longitude)[0].toString())
+            } catch (e: Exception) {
+                logTag("SCOOTER_ERROR", e.toString())
```
Show an error if the API call fails

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapViewModel.kt:
```diff
@@ -0,0 +1,24 @@
+        viewModelScope.launch {
+            try {
+                scooterList.value = mapService.getAllScooters(latitude = latitude, longitude = longitude)
+                logTag("SCOOTER_0", mapService.getAllScooters(latitude = latitude, longitude = longitude)[0].toString())
```
`mapService.getAllScooters(latitude = latitude, longitude = longitude)[0]` will crash if the coordinates list is empty.
Use `.getOrNull(0)` instead

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/splash/SplashFragment.kt:
```diff
@@ -27,7 +27,10 @@ class SplashFragment : Fragment(R.layout.fragment_splash) {
                 viewModel.getAuthData()
                 if (viewModel.userData.value != null) {
-                    findNavController().navigate(SplashFragmentDirections.actionGlobalMapFragment())
+                    if (viewModel.userData.value!!.user.driverLicenseKey.isNullOrEmpty()) {
```
Avoid using `!!`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/splash/SplashFragment.kt:
```diff
@@ -27,7 +27,10 @@ class SplashFragment : Fragment(R.layout.fragment_splash) {
                 if (viewModel.userData.value != null) {
-                    findNavController().navigate(SplashFragmentDirections.actionGlobalMapFragment())
+                    if (viewModel.userData.value!!.user.driverLicenseKey.isNullOrEmpty()) {
+                        findNavController().navigate(SplashFragmentDirections.actionGlobalLicenseInstructionsFragment(viewModel.userData.value!!))
```
Avoid using `!!`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/SessionInterceptor.kt:
```diff
@@ -0,0 +1,27 @@
+
+class SessionInterceptor(
+    private val authenticationTokenProvider: AuthenticationTokenProvider
+//    private val internalStorageManager: UserDataInternalStorageManager
```
Remove unused commented code

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/OnBoardingRepository.kt:
```diff
@@ -8,7 +8,5 @@ class OnBoardingRepository(private val userDataInternalStorageManager: UserDataI
-    suspend fun getAuthStatus(): UserResponse? {
-        return userDataInternalStorageManager.getAuthPreferences()
-    }
+    suspend fun getAuthStatus(): UserResponse? = userDataInternalStorageManager.getAuthPreferences()
```
`suspend` modifier seems to be redundant for these methods

catadarjan-tapptitude replied:
I know the `UserResponse` class has been in the code for a while now, but I think it would be appropriate to rename it using the `Dto` suffix. This is because you're using it as a Network API class

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/OnBoardingRepository.kt:
```diff
@@ -8,7 +8,5 @@ class OnBoardingRepository(private val userDataInternalStorageManager: UserDataI
 
     suspend fun changeLoggedStatus(logValue: Boolean) = userDataInternalStorageManager.changeLogStatus(logValue)
```
`logValue` could be renamed to `isLoggedIn` to be more clear
Same would go for `changeLogStatus`, to be `setIsUserLoggedIn(..)`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/licenseRegistration/LicenseRegistrationViewModel.kt:
```diff
@@ -35,14 +35,13 @@ class LicenseRegistrationViewModel(
+                _isError.value = true
             } finally {
                 isLoading.value = false
+                _isError.value = false
```
You're overwritting the `isError` value here. 
Since you set it in `catch` and in `finally` you're triggering the observer twice needlessly

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/SessionInterceptor.kt:
```diff
@@ -0,0 +1,27 @@
+    private val authenticationTokenProvider: AuthenticationTokenProvider
+//    private val internalStorageManager: UserDataInternalStorageManager
+) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
```
Leave an empty line between class declaration and first method/field.

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/SessionInterceptor.kt:
```diff
@@ -0,0 +1,27 @@
+        val url = chain.request().url.toString()
+        val token = authenticationTokenProvider.getAuthToken()
+
+        if (token != null && !url.contains("login") && !url.contains("register")) {
```
"login" and "register" could be extracted to static constants in the companion object. Also, as mentioned in `LicenseRegistrationViewModel` comment, you could add a cheeck for the license verificationm api call

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/UserDataInternalStorageManager.kt:
```diff
@@ -3,45 +3,45 @@ package com.internship.move.model
+import com.squareup.moshi.Moshi
 
-class UserDataInternalStorageManager(context: Context) {
+class UserDataInternalStorageManager(context: Context, private val moshi: Moshi) {
```
There are a few things to talk about in this class. Let's discuss in a huddle

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/appModule.kt:
```diff
@@ -18,29 +23,35 @@ val viewModel = module {
 
 fun provideLicenseService(retrofit: Retrofit): LicenseService = retrofit.create(LicenseService::class.java)
 
+fun provideMapService(retrofit: Retrofit): MapService = retrofit.create(MapService::class.java)
```
These provider methods should be private

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/providers/AuthenticationTokenProvider.kt:
```diff
@@ -0,0 +1,5 @@
+package com.internship.move.model.providers
+
+interface AuthenticationTokenProvider {
+    fun getAuthToken(): String?
```
Leave an empty line between class declaration and first method/field

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/model/providers/RuntimeAuthenticationTokenProvider.kt:
```diff
@@ -0,0 +1,7 @@
+import com.internship.move.model.UserDataInternalStorageManager
+
+class RuntimeAuthenticationTokenProvider(private val internalStorageManager: UserDataInternalStorageManager) : AuthenticationTokenProvider {
+    override fun getAuthToken() = internalStorageManager.getUserToken()
```
Leave an empty line between class declaration and first method/field

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/utils/Utils.kt:
```diff
@@ -53,4 +53,6 @@ fun checkUserOrPassword(userOrPasswordText: String?): Boolean {
-const val ERROR = -1
\ No newline at end of file
+const val ERROR = -1
+const val MAPS_API_KEY = "AIzaSyDjK7n4r_ENUgIKBNIsxMD8J09ADCOVUGk"
```
This constant seems to be unused

catadarjan-tapptitude commented on app/src/main/res/color/edit_text_box_stroke.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:color="@color/neutral_white" android:state_focused="true" />
```
Create specific color values instead of referencing the the base color.
That way you'll only need to set the right color value for dark more colors.xml instead of creating a new selector

catadarjan-tapptitude commented on app/src/main/res/color/edit_text_box_stroke.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I'm leaving this comment here because I can't comment on `.png` files:
`launcher.png` and `scooter_img.png` should be renamed to respect the naming conventions

catadarjan-tapptitude commented on app/src/main/res/drawable/bg_small_icon_btn.xml:
```diff
@@ -0,0 +1,12 @@
+
+    <solid android:color="@color/accent_pink" />
+
+    <corners android:radius="14dp" />
```
Extract raw values

catadarjan-tapptitude commented on app/src/main/res/drawable/bg_small_icon_btn.xml:
```diff
@@ -0,0 +1,12 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+
+    <solid android:color="@color/accent_pink" />
```
Specific colors

catadarjan-tapptitude commented on app/src/main/res/drawable/color_cursor.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <size android:width="2dp" />
```
Extract raw values

catadarjan-tapptitude commented on app/src/main/res/drawable/color_cursor.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <size android:width="2dp" />
+    <solid android:color="@color/neutral_white" />
```
Specific colors

catadarjan-tapptitude commented on app/src/main/res/drawable/icon_tint_selector.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:color="?attr/colorAccent" android:state_focused="true" />
```
This file appears to not be used anywhere

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_license_confirm.xml:
```diff
@@ -27,13 +27,28 @@
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toBottomOf="@+id/imageView"
-        tools:text="We are validating your license" />
+        app:layout_constraintTop_toBottomOf="@+id/licenseConfirmIV"
```
Use `@id/` when reffering to existing ids. Use `@+id/` only when declaring a new id

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="utf-8"?>
+
```
Extra empty line.
Also, is this a fragment?

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/scooterCV"
+    android:layout_width="300dp"
```
Extract raw values

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+            android:layout_height="wrap_content"
+            android:layout_marginTop="4dp"
+            android:layout_marginRight="@dimen/textView_secondary_margin_end"
+            android:textColor="@color/primary_dark_purple"
```
Specific colors

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/textView_margin_top"
+            android:layout_marginRight="@dimen/textView_secondary_margin_end"
```
Don't use view type specific names in res names. Unless it's something like a "button" which is a bit more generic

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/scooterLocationTV"
+            tools:text="Unlock" />
+
```
Extra empty lines

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_map_scooter_item.xml:
```diff
@@ -0,0 +1,150 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/scooterLocationTV"
```
`@id/`

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_onboarding.xml:
```diff
@@ -8,41 +8,38 @@
-        android:layout_marginStart="24dp"
-        android:layout_marginBottom="98dp"
+        android:layout_marginStart="@dimen/dots_indicator_margin_start"
+        android:layout_marginBottom="58dp"
```
Extract raw dimensions
(check other files where this applies)

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_onboarding.xml:
```diff
@@ -8,41 +8,38 @@
+        app:dotsSize="8dp"
         app:dotsSpacing="4dp"
-        app:dotsStrokeColor="@color/neutral_grey"
+        app:dotsStrokeColor="@color/primary_dark_purple"
```
Specific colors
(check other files where this applies)

catadarjan-tapptitude commented on app/src/main/res/layout/fragment_onboarding_page.xml:
```diff
@@ -34,14 +35,17 @@
+    <Button
         android:id="@+id/skipTV"
-        style="@style/RegularText"
+        style="@style/skipButtonStyle"
```
"SkipButtonStyle" is actually a "TextButtonStyle".
You can use this style for other buttons that look similarly and can be named more generically

catadarjan-tapptitude commented on app/src/main/res/values/strings.xml:
```diff
@@ -14,22 +16,25 @@
     <string name="text_button_onboarding_page">Next</string>
 
-    <!--    LOGIN TO REGISTER  -->
+    <!--    LOGIN TO REGISTER STRINGS  -->
```
Same thing about region comments as in gradle
`<!-- region Authentication -->` combined with `<!-- endregio
![image](https://user-images.githubusercontent.com/84084708/191179579-80c973c7-4c98-4e6e-9bb7-dce7a4f94370.png)
![image](https://user-images.githubusercontent.com/84084708/191179595-c499507d-8fc3-45e6-85f9-0f5eb956279e.png)

n -->` will allow you to collapse regions

catadarjan-tapptitude commented on app/src/main/res/values/strings.xml:
```diff
@@ -14,22 +16,25 @@
 
-    <!--    LOGIN TO REGISTER  -->
+    <!--    LOGIN TO REGISTER STRINGS  -->
     <string name="goToRegisterText">Don\'t have an account? You can</string>
```
Use snake_case for resource names

catadarjan-tapptitude commented on build.gradle:
```diff
@@ -3,16 +3,20 @@
     id 'com.android.library' version '7.2.1' apply false
     id 'org.jetbrains.kotlin.android' version '1.7.10' apply false
+    id 'com.google.android.libraries.mapsplatform.secrets-gradle-plugin' version '2.0.1' apply false
+
```
Extra empty line

catadarjan-tapptitude commented on build.gradle:
```diff
@@ -3,16 +3,20 @@
     dependencies {
         classpath("androidx.navigation:navigation-safe-args-gradle-plugin:2.5.1")
     }
+
```
Extra empty line

catadarjan-tapptitude commented on build.gradle:
```diff
@@ -3,16 +3,20 @@
         google()
+        maven { url "https://jitpack.io" }
     }
     dependencies {
```
Could leave an empty line between these 2 blocks

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
-class MapFragment : Fragment(R.layout.fragment_map) {
-
+class MapFragment : Fragment(R.layout.fragment_map), OnMapReadyCallback {
     private val binding by viewBinding(FragmentMapBinding::bind)
```
Leave an empty line between class declaration and first method/field.

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+        }
+    }
+
+    private fun bitmapDescriptorFromVector(vectorResId: Int): BitmapDescriptor {
```
I would bet that you copy pasted this method from Java and Android Studio transformed it to kotlin. Am I right? 😁 
This isn't a problem, we do it all the time. But, Android Studio kotlin to java transformation isn't perfect and many times leaves us with lint warnings that should be fixed:
`vectorDrawable.getIntrinsicWidth()` should be replaced with `vectorDrawable.intrinsicWidth` (similar for height)
Replace `!!` calls with `?.` or null checks

catadarjan-tapptitude replied:
Use `ResourcesCompat.getDrawable(..)` instead of `ContextCompat` to avoid crashes

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+    private fun bitmapDescriptorFromVector(vectorResId: Int): BitmapDescriptor {
+        val vectorDrawable = ContextCompat.getDrawable(requireContext(), vectorResId)
+        vectorDrawable!!.setBounds(0, 0, vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight())
+        val bitmap = Bitmap.createBitmap(vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888)
```
Directly import `ARGB_8888`

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+    private fun initButtons() {
+        requireActivity().onBackPressedDispatcher.addCallback {
+            if (doubleBackPressed) {
+                authViewModel.logOut()
```
Why are you logging out the user when closing the app?

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+            } else {
+                doubleBackPressed = true
+                Toast.makeText(requireContext(), getString(R.string.exit_app_button_text), Toast.LENGTH_SHORT).show()
+                Handler().postDelayed({
```
The compiler is telling you that you're using a deprecated constructor.
In this case it can be fixed using `Handler(Looper.getMainLooper())`
Please fix compiler lint warnings whenever possible

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+                Toast.makeText(requireContext(), getString(R.string.exit_app_button_text), Toast.LENGTH_SHORT).show()
+                Handler().postDelayed({
+                    doubleBackPressed = false
+                }, 3000L)
```
Extract magic numbers to static constants

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
-            authViewModel.logOut()
+    private fun initScooterInfo(markerTitle: String) {
+        mapViewModel.scooterList.value?.forEach { scooter ->
+            if (scooter._id == markerTitle) {
```
The `_id` field can be renamed to `id`. You're using `@Json(name = "_id")` to tell Moshi how to map it. This allows you to name the fields however you need

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
-            findNavController().navigate(MapFragmentDirections.actionMapFragmentToNavigationIntro())
+    private fun setBatteryIcon(batteryLevel: String) {
+        val resource = when (batteryLevel.toInt()) {
+            in 81..100 -> R.drawable.ic_battery_100
```
Instead of using `in` you could refactor this method a bit to:
```
    private fun setBatteryIcon(batteryLevel: Int) {
        val resource = when {
            batteryLevel >= 81 -> R.drawable.ic_battery_100
            batteryLevel >= 61 -> R.drawable.ic_battery_80
            // etc ..
            else -> R.drawable.ic_battery_0
        }
        binding.scooterInfo.batteryIndicatorIV.setImageResource(resource)
    }
```

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+    private fun checkPermissions(): Boolean {
+        if (ActivityCompat.checkSelfPermission(
+                requireContext(),
+                Manifest.permission.ACCESS_COARSE_LOCATION
```
SInce fine location permissions comes with coarse location, it's enough to check and ask for fine location permissions

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+    }
+
+    @SuppressLint("MissingSuperCall")
+    override fun onRequestPermissionsResult(
```
This method is deprecated. Use contracts instead, as mentioned in the provided documentation
https://developer.android.com/training/location/permissions#user-choice-affects-permission-grants

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
+            scooterCoords.coordinates[0],
+            1
+        ).get(0).thoroughfare
+            .toString() + ", Nr. " + Geocoder(requireContext(), Locale.getDefault()).getFromLocation(
```
Format this using strings

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
 import com.internship.move.utils.logTag
 import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
 import org.koin.androidx.viewmodel.ext.android.viewModel
+import java.util.*
```
Avoid using * imports

catadarjan-tapptitude commented on app/src/main/java/com/internship/move/feature/map/MapFragment.kt:
```diff
@@ -1,47 +1,261 @@
 package com.internship.move.feature.map
```
We can talk more about this file after you finish solving the current comments 😁

IanisT227 committed:
Implemented most of the requirements from PR
Changed the way Image is being transmitted from a fragment to another
renamed and added some colors, dimens and strings

IanisT227 committed:
Tried to implement cluster manager
changed some logic according to the previous PR

IanisT227 committed:
Solved clustering and added bottom sheet dialog

IanisT227 committed:
Extracted dimensions and strings for bottom_sheet_scooter_card.xml

IanisT227 committed:
Map fragment styling and move to live location button
almost visually implemented unlock by scooter number
